### PR TITLE
ENH: f2py: improve error messages

### PR DESCRIPTION
**Pull request by eric-wieser**
I just got this error while hacking some scipy internals (https://github.com/scipy/scipy/pull/11158), and had nowhere near enough information to debug it.

This is a little closer to useful.

### CODE REVIEW

File: numpy/f2py/rules.py [hunk 1]
```diff
@@ -1081,8 +1083,10 @@
 \t\t\twhile ((_i = nextforcomb()))
 \t\t\t\t#varname#[capi_i++] = #init#; /* fortran way */
 \t\t} else {
-\t\t\tif (!PyErr_Occurred())
-\t\t\t\tPyErr_SetString(#modulename#_error,\"Initialization of #nth# #varname# failed (initforcomb).\");
+\t\t\tPyObject *exc, *val, *tb;
+\t\t\tPyErr_Fetch(&exc, &val, &tb);
+\t\t\tPyErr_SetString(#modulename#_error,\"Initialization of #nth# #varname# failed (initforcomb).\");
+\t\t\tnpy_PyErr_ChainExceptionsCause(exc, val, tb);
```
Comment (eric-wieser) [inline numpy/f2py/rules.py]: Possible I have a missing include here, wasn't quite able to work out how to add `npy_3kcompat.h` to the include list. Let's see if its already there...

File: numpy/f2py/src/fortranobject.c [hunk 2]
```diff
@@ -839,9 +839,9 @@ PyArrayObject* array_from_pyobj(const int type_num,
     if ((intent & F2PY_INTENT_INOUT) ||
             (intent & F2PY_INTENT_INPLACE) ||
             (intent & F2PY_INTENT_CACHE)) {
-        PyErr_SetString(PyExc_TypeError,
-                        "failed to initialize intent(inout|inplace|cache) "
-                        "array, input not an array");
+        PyErr_Format(PyExc_TypeError,
+                     "failed to initialize intent(inout|inplace|cache) "
+                     "array, input '%s' object is not an array", Py_TYPE(obj)->tp_name);
```
Comment (WarrenWeckesser) [inline numpy/f2py/src/fortranobject.c]: Nit: according to the [NumPy C Style Guide](https://github.com/numpy/numpy/blob/master/doc/C_STYLE_GUIDE.rst.txt), lines should not be more than 79 characters long.

Comment (eric-wieser) [reply numpy/f2py/src/fortranobject.c]: This file is full of violations, but will fix if this needs another revision to pass CI.

File: numpy/f2py/rules.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/f2py/rules.py]: `ImportError: /tmp/tmp4n87gp4z/_test_ext_module_5427.cpython-36m-i386-linux-gnu.so: undefined symbol: npy_PyErr_ChainExceptionsCause`

File: numpy/f2py/src/fortranobject.c (hunk 2 above)
Comment (WarrenWeckesser) [reply numpy/f2py/src/fortranobject.c]: Yes, but shouldn't we strive to not make it worse?

Comment (eric-wieser) [reply numpy/f2py/src/fortranobject.c]: It did, so fixed :)

Comment (eric-wieser) [reply numpy/f2py/src/fortranobject.c]: @WarrenWeckesser: You could argue that a ~.5% increase in style quality in this file is not worth the power consumption of a full CI run. However, given CI failed that became irrelevant.

**Comment by eric-wieser** [comment]
Test failures are real, because this has changed exceptions like
```
ValueError: failed to initialize intent(inout) array -- input not fortran contiguous
```
to
```
ValueError: failed to initialize intent(inout) array -- input not fortran contiguous

The above exception was the direct cause of the following exception:

_test_ext_module_5413.error: failed in converting 1st argument `x' of _test_ext_module_5413.foo_no to C/Fortran array
```
which means they can no longer be caught as `ValueError`.

Is this ok? The error is far more useful now.

**Comment by eric-wieser** [comment]
Went for option 3 from my list above, since it was easy - all tests pass

File: numpy/f2py/rules.py (hunk 1 above)
Comment (seberg) [reply numpy/f2py/rules.py]: I am a bit confused on this. Aren't these new models generated by f2py. And in that case they cannot rely on (private?) numpy API that is not available in older numpy versions?

Comment (seberg) [reply numpy/f2py/rules.py]: Hmm, but then... it works, is fortranobject.so shipped? Have to dig a bit I guess. And as for version, we have backward compatibility, not forward compatibility, so its OK if someone wants to ship compiled modules that they have to compile with the oldest f2py/numpy.

Comment (eric-wieser) [reply numpy/f2py/rules.py]: My very lacking understanding was that f2py was a build tool, and in practice all that's ever shipped is a script that invokes it, or a binary built with it - in both cases, all the needed functions are inlined or public API.

Comment (eric-wieser) [reply numpy/f2py/rules.py]: What you're perhaps missing is that fortranobjecr.h is included by the generated code, which is why I added the include there.

**Comment by mattip** [comment]
This also reduces the repeated compatibility shims, so worthwhile as a cleanup too.
Merging.

**Comment by mattip** [comment]
Thanks @eric-wieser

**Comment by eric-wieser** [comment]
Let's hope that [@seberg's concern](https://github.com/numpy/numpy/pull/15022#discussion_r352820609) turns out to be a non-issue.

**Pull request merged**